### PR TITLE
feat(ingester): Respect `IngestState` during WAL replay

### LINKED ISSUE #8615: Respect ingest system state during WAL replay
In the ingester we have a system to signal saturation (of various things) and apply back pressure by rejecting writes.

During WAL replay, we don't listen to this system - lets make the WAL replay check the `IngestState` signals from other sub-systems, and stop replay when it signals bad things.

### PR DESCRIPTION
**Pull request by savage-engineer**
Fairly small change with the previous addition of `IngestState::read_with_exceptions()`, so this is mostly tests and visibility shenanigans. `replay()` is benchmarked and `IngestState` was not exported. I've done the good old feature conditional `pub mod`/`mod`, so `wal_replay` is only public when the benches feature is enabled.

Closes #8615

---

* refactor(ingester): Pass `IngestState` to WAL replay (8548ea3b3)

      This requires the `IngestState` and associated types to be public so
      that WAL replay can be called by the benchmarker. The module containing
      the `IngestState` is private and is only conditionally re-exported under
      the benchmark feature as part of the `internal_implementation_details`
      module.

* feat(ingester): Respect non-disk full `IngestStateError` during WAL replay (9e80e0306)

      This change causes WAL replay to mimic the RPC write handler, mostly
      respecting the `IngestState` before apply an op while replaying a WAL
      file. The caveat is that `DiskFull` is ignored as WAL replay specifically
      helps with this state

* test(ingester): Add timeout panic to blocked `IngestState` WAL replay test (f6d6dd9b5)

### CODE REVIEW

File: ingester/src/init/wal_replay.rs [hunk 1]
```diff
@@ -349,6 +364,20 @@ where
...
+                    Err(e) => {
+                        warn!(
+                            ingest_state_error=%e,
+                            wait_duration=?OP_REPLAY_BACKPRESSURE_WAIT_DURATION,
+                            "ingest state is unhealthy, waiting for ingest state to recover before replaying wal op",
+                        );
+                        tokio::time::sleep(OP_REPLAY_BACKPRESSURE_WAIT_DURATION).await;
+                    }
+                }
+            }
```
Comment (savage-engineer) [inline ingester/src/init/wal_replay.rs:379]: This is the functional change. I had considered passing down a `backoff::BackoffConfig` but not sure if it adds much.

Comment (savage-engineer) [reply ingester/src/init/wal_replay.rs:379]: The state is checked before each call to `apply()` on the sink, to mirror the RPC write path.

File: ingester/src/ingest_state.rs [hunk 2]
```diff
@@ -14,13 +14,18 @@ use thiserror::Error;
...
+    /// persist jobs.
     #[error("ingester overloaded - persisting backlog")]
     PersistSaturated = 1 << 0,
 
+    /// Set for the duration of the ingester's graceful shutdown procedure.
     #[error("ingester is shutting down")]
     GracefulStop = 1 << 1,
 
+    /// Indicates the ingester's disk is too full to safely accept further
+    /// writes.
```
Comment (domodwyer) [inline ingester/src/ingest_state.rs:28]: :+1:

File: ingester/src/init/wal_replay.rs (hunk 1 above)
Comment (domodwyer) [reply ingester/src/init/wal_replay.rs:379]: Yeah backing off doesn't add any value, but would slow everything down when it grows to >30s checks - good call :+1:

File: ingester/src/ingest_state.rs [hunk 3]
```diff
@@ -64,7 +69,7 @@ impl IngestState {
     ///
     /// Returns true if this call set the error state to `error`, false if
     /// `error` was already set.
-    pub(crate) fn set(&self, error: IngestStateError) -> bool {
+    pub fn set(&self, error: IngestStateError) -> bool {
```
Comment (domodwyer) [inline ingester/src/ingest_state.rs:72]: Does `set()` and `unset()` need to be in the public scope? Only the ingester should ever be changing the state

Comment (savage-engineer) [reply ingester/src/ingest_state.rs:72]: That's a good point - I don't think any of the methods actually need to be public outside the crate. I'll take a stab at reverting them back.

**Pull request merged**